mlem/contrib/venv.py

```diff
@@ -0,0 +1,173 @@
+
+    def create_virtual_env(self):
+        env_spec = venv.EnvBuilder(with_pip=True)
+        env_dir = os.path.abspath(self.target)
+
+        context = env_spec.ensure_directories(env_dir)
+
+        true_system_site_packages = env_spec.system_site_packages
+        env_spec.system_site_packages = False
+        env_spec.create_configuration(context)
+        env_spec.setup_python(context)
+        if env_spec.with_pip:
+            env_spec._setup_pip(context)  # pylint: disable=protected-access
+        if not env_spec.upgrade:
+            env_spec.setup_scripts(context)
+            env_spec.post_setup(context)
+        if true_system_site_packages:
+            env_spec.system_site_packages = True
+            env_spec.create_configuration(context)
+
+        return context
+
+    def get_installed_packages(self, env_dir):
+        if platform.system() == "Windows":
+            env_exe = os.path.join(env_dir, "Scripts", "python.exe")
+        else:
+            env_exe = os.path.join(env_dir, "bin", "python")
+
+        return subprocess.check_output([env_exe, "-m", "pip", "freeze"])
+
+    def build(self, obj: MlemModel):
+        if self.current_env:
+            if (
+                os.getenv("VIRTUAL_ENV") is None
+                or sys.prefix == sys.base_prefix
+            ):
+                raise MlemError("No virtual environment detected.")
+            echo(EMOJI_PACK + f"Detected the virtual env {sys.prefix}")
+            env_dir = sys.prefix
+            env_exe = os.path.join(env_dir, "bin", "python")
+        else:
+            echo(EMOJI_PACK + f"Creating virtual env {self.target}...")
+            context = self.create_virtual_env()
+            env_dir = context.env_dir
+            os.environ["VIRTUAL_ENV"] = context.env_dir
+            env_exe = context.env_exe
+        echo(EMOJI_PACK + "Installing the required packages...")
+        # Based on recommendation given in https://pip.pypa.io/en/latest/user_guide/#using-pip-from-your-program
+        install_cmd = [env_exe, "-m", "pip", "install"]
+        if self.no_cache:
+            install_cmd.append("--no-cache-dir")
+        install_cmd.extend(obj.requirements.to_pip())
+        subprocess.run(install_cmd, check=True)
+        echo(
+            EMOJI_OK
+            + f"virtual environment `{self.target}` is ready, activate with `source {self.target}/bin/activate`"
+        )
+        return env_dir
+
+
+class CondaBuilder(EnvBuilder):
+
+    type: ClassVar = "conda"
+
+    python_version: str = "3"
+    """The python version to use"""
+    current_env: Optional[bool] = False
+    """Whether to install in the current conda env"""
+
+    def create_virtual_env(self):
+        ret = None
+        create_cmd = ["--prefix", self.target, f"python={self.python_version}"]
+        try:
+            ret = subprocess.run(
+                ["conda", "create", "-y", *create_cmd], check=True
+            )
+        except FileNotFoundError:
```

Comment:
should we just `os.makedirs` instead?

Reply:
I don't think so, because `conda` create does more than just making a directory.